File: packages/react-native/src/client-internal/content-manager.ts
import type { PolycentricClient } from '../polycentric-client';
import { polycentric as proto } from '../generated/protocol';

export class ContentManager {
  constructor(private readonly client: PolycentricClient) {}

  private _system(): proto.IPublicKey {
    return this.client.currentKeyPair.publicKey;
  }

  private _process(): proto.IProcess {
    return this.client.process;
  }

  private async _createLWWElementSetEvent(
    contentType: proto.ContentType,
    value: Uint8Array,
    operation: proto.LWWElementSet.Operation
  ): Promise<proto.SignedEvent> {
    const lwwElementSet = proto.LWWElementSet.create({
      operation,
      value,
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType,
      lwwElementSet,
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  private async _createOpinion(
    opinion: proto.Opinion,
    subjectPointer: proto.IPointer
  ): Promise<proto.SignedEvent> {
    const subjectReference = proto.Reference.create({
      referenceType: 2, // TODO: Create Proto ReferenceType enum, reference type of 2 is a Pointer
      reference: proto.Pointer.encode(subjectPointer).finish(),
    });

    const lwwElement = proto.LWWElement.create({
      value: new Uint8Array([opinion]),
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.OPINION,
      lwwElement,
      references: [subjectReference],
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  private async _createDelete(
    targetPointer: proto.IPointer,
    contentType: proto.ContentType
  ): Promise<proto.SignedEvent> {
    const deleteEvent = proto.Delete.create({
      process: targetPointer.process,
      logicalClock: targetPointer.logicalClock,
      unixMilliseconds: Date.now(),
      indices: proto.Indices.create({}),
      contentType,
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.DELETE,
      content: proto.Delete.encode(deleteEvent).finish(),
      references: [
        proto.Reference.create({
          referenceType: 2,
          reference: proto.Pointer.encode(targetPointer).finish(),
        }),
      ],
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async _createEvent(
    eventData: proto.EventCreationData
  ): Promise<proto.SignedEvent> {
    if (!this.client.hasIdentity()) {
      const error = new Error(
        'Cannot create events without an active identity'
      );
      this.client.events.emitError(error);
      throw error;
    }

    eventData.logicalClock = this.client.nextLogicalClock();
    eventData.unixMilliseconds = Date.now();

    // createEvent returns raw unsigned Event bytes for signing
    const eventBytes = this.client.ffiBridge.createEvent(eventData, Date.now());

    // Sign the event bytes via CryptoManager
    const signature = await this.client.cryptoManager.sign(
      this.client.currentKeyPair.privateKey.key,
      eventBytes,
      this.client.currentKeyPair.keyType
    );

    // Wrap in SignedEvent protobuf
    const signedEvent = proto.SignedEvent.create({
      signature,
      event: eventBytes,
    });

    // Ingest into the Rust core's local store
    this.client.ffiBridge.ingestEvent(signedEvent);

    // Persist event to local database
    if (this.client.storage) {
      this.client.storage.events.persistEvent(signedEvent);
    }

    this.client.events.emitContentCreated(signedEvent);

    return signedEvent;
  }

  async createPost(
    content: string,
    image?: proto.IImageManifest,
    reference?: proto.IReference
  ): Promise<proto.SignedEvent> {
    const post = proto.Post.create({ content, image });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.POST,
      content: proto.Post.encode(post).finish(),
      references: reference ? [reference] : [],
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async createLike(subjectPointer: proto.IPointer): Promise<proto.SignedEvent> {
    return this._createOpinion(proto.Opinion.LIKE, subjectPointer);
  }

  async createDislike(
    subjectPointer: proto.IPointer
  ): Promise<proto.SignedEvent> {
    return this._createOpinion(proto.Opinion.DISLIKE, subjectPointer);
  }

  async createNeutral(
    subjectPointer: proto.IPointer
  ): Promise<proto.SignedEvent> {
    return this._createOpinion(proto.Opinion.NEUTRAL, subjectPointer);
  }

  async createUsername(username: string): Promise<proto.SignedEvent> {
    const lwwElement = proto.LWWElement.create({
      value: new TextEncoder().encode(username),
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.USERNAME,
      lwwElement,
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async createDescription(description: string): Promise<proto.SignedEvent> {
    const lwwElement = proto.LWWElement.create({
      value: new TextEncoder().encode(description),
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.DESCRIPTION,
      lwwElement,
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async createAvatar(avatar: proto.IImageManifest): Promise<proto.SignedEvent> {
    const lwwElement = proto.LWWElement.create({
      value: proto.ImageManifest.encode(avatar).finish(),
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.AVATAR,
      lwwElement,
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async createBanner(banner: proto.IImageManifest): Promise<proto.SignedEvent> {
    const lwwElement = proto.LWWElement.create({
      value: proto.ImageManifest.encode(banner).finish(),
      unixMilliseconds: Date.now(),
    });

    const eventData = proto.EventCreationData.create({
      contentType: proto.ContentType.BANNER,
      lwwElement,
      system: this._system(),
      process: this._process(),
    });

    return this._createEvent(eventData);
  }

  async createFollow(system: proto.IPublicKey): Promise<proto.SignedEvent> {
    const systemBytes = proto.PublicKey.encode(system).finish();
    return this._createLWWElementSetEvent(
      proto.ContentType.FOLLOW,
      systemBytes,
      proto.LWWElementSet.Operation.ADD
    );
  }

  async createUnfollow(system: proto.IPublicKey): Promise<proto.SignedEvent> {
    const systemBytes = proto.PublicKey.encode(system).finish();
    return this._createLWWElementSetEvent(
      proto.ContentType.FOLLOW,
      systemBytes,
      proto.LWWElementSet.Operation.REMOVE
    );
  }

  async createBlock(system: proto.IPublicKey): Promise<proto.SignedEvent> {
    const systemBytes = proto.PublicKey.encode(system).finish();
    return this._createLWWElementSetEvent(
      proto.ContentType.BLOCK,
      systemBytes,
      proto.LWWElementSet.Operation.ADD
    );
  }

  async createUnblock(system: proto.IPublicKey): Promise<proto.SignedEvent> {
    const systemBytes = proto.PublicKey.encode(system).finish();
    return this._createLWWElementSetEvent(
      proto.ContentType.BLOCK,
      systemBytes,
      proto.LWWElementSet.Operation.REMOVE
    );
  }

  async createAddServer(server: string): Promise<proto.SignedEvent> {
    const serverBytes = new TextEncoder().encode(server);
    return this._createLWWElementSetEvent(
      proto.ContentType.SERVER,
      serverBytes,
      proto.LWWElementSet.Operation.ADD
    );
  }

  async createRemoveServer(server: string): Promise<proto.SignedEvent> {
    const serverBytes = new TextEncoder().encode(server);
    return this._createLWWElementSetEvent(
      proto.ContentType.SERVER,
      serverBytes,
      proto.LWWElementSet.Operation.REMOVE
    );
  }

  async createAddAuthority(authority: string): Promise<proto.SignedEvent> {
    const authorityBytes = new TextEncoder().encode(authority);
    return this._createLWWElementSetEvent(
      proto.ContentType.AUTHORITY,
      authorityBytes,
      proto.LWWElementSet.Operation.ADD
    );
  }

  async createRemoveAuthority(authority: string): Promise<proto.SignedEvent> {
    const authorityBytes = new TextEncoder().encode(authority);
    return this._createLWWElementSetEvent(
      proto.ContentType.AUTHORITY,
      authorityBytes,
      proto.LWWElementSet.Operation.REMOVE
    );
  }

  async createJoinTopic(topic: string): Promise<proto.SignedEvent> {
    const topicBytes = new TextEncoder().encode(topic);
    return this._createLWWElementSetEvent(
      proto.ContentType.JOIN_TOPIC,
      topicBytes,
      proto.LWWElementSet.Operation.ADD
    );
  }

  async createLeaveTopic(topic: string): Promise<proto.SignedEvent> {
    const topicBytes = new TextEncoder().encode(topic);
    return this._createLWWElementSetEvent(
      proto.ContentType.JOIN_TOPIC,
      topicBytes,
      proto.LWWElementSet.Operation.REMOVE
    );
  }

  async setOpinion(
    pointer: proto.IPointer,
    opinion: proto.Opinion
  ): Promise<proto.SignedEvent> {
    return this._createOpinion(opinion, pointer);
  }

  async deletePost(postPointer: proto.IPointer): Promise<proto.SignedEvent> {
    return this._createDelete(postPointer, proto.ContentType.POST);
  }
}

File: packages/kotlin/app/src/main/java/tech/futo/libPolycentric/services/ContentManager.kt
package tech.futo.libPolycentric.services

import PolycentricException
import okio.ByteString
import okio.ByteString.Companion.toByteString
import polycentric.*
import tech.futo.libPolycentric.PolycentricClient

class ContentManager(private val client: PolycentricClient) {
    private suspend fun createLWWElementSetEvent(
        contentType: ContentType,
        value: ByteArray,
        operation: LWWElementSet.Operation,
    ): SignedEvent {
        val lwwElementSet = LWWElementSet(
            operation = operation,
            value_ = value.toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = contentType,
            lww_element_set = lwwElementSet,
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    private suspend fun getReference(pointer: Pointer): EventKey? {
        val pointerBytes = Pointer.ADAPTER.encode(pointer)
        val result = this.client.ffiService.getReference(pointerBytes)
        if (result.isEmpty()) return null
        return EventKey.ADAPTER.decode(result)
    }

    private suspend fun createOpinion(opinion: Opinion, subjectPointer: Pointer): SignedEvent {
        getReference(subjectPointer)
            ?: throw PolycentricException("Could not get reference from pointer")

        val subjectReference = Reference(
            reference_type = 2L,
            reference = Pointer.ADAPTER.encode(subjectPointer).toByteString(),
        )

        val lwwElement = LWWElement(
            value_ = byteArrayOf(opinion.value.toByte()).toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.OPINION,
            lww_element = lwwElement,
            references = listOf(subjectReference),
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    private suspend fun createDelete(targetPointer: Pointer, contentType: ContentType): SignedEvent {
        val deleteEvent = Delete(
            process = targetPointer.process,
            logical_clock = targetPointer.logical_clock,
            unix_milliseconds = System.currentTimeMillis(),
            indices = Indices(),
            content_type = contentType,
        )

        val eventData = EventCreationData(
            content_type = ContentType.DELETE,
            content = Delete.ADAPTER.encode(deleteEvent).toByteString(),
            references = listOf(
                Reference(
                    reference_type = 2L,
                    reference = Pointer.ADAPTER.encode(targetPointer).toByteString(),
                )
            ),
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createEvent(eventData: EventCreationData): SignedEvent {
        val identity = client.currentIdentity
        val processBytes = identity.process.process.toByteArray()
        val systemKeyType = identity.keyPair.keyType
        val systemKeyBytes = identity.keyPair.publicKey.key.toByteArray()

        val logicalClock = client.processStateRepository.getNextLogicalClock(
            systemKeyType,
            systemKeyBytes,
            processBytes,
        )

        val eventDataWithClock = eventData.copy(logical_clock = logicalClock)
        val eventDataBytes = EventCreationData.ADAPTER.encode(eventDataWithClock)
        val unixMs = System.currentTimeMillis().toInt()

        val event = client.ffiService.createEvent(eventDataBytes, unixMs)
        val signature = client.identityManager.sign(event.toByteString())

        val signedEvent = SignedEvent(event = event.toByteString(), signature = signature)

        client.ffiService.ingestEvent(signedEvent.encode())

        client.eventRepository.persistEvent(signedEvent)
        client.processStateRepository.persistCurrentLogicalClock(
            systemKeyType,
            systemKeyBytes,
            processBytes,
            logicalClock,
        )

        client.events.emitContentCreated(Event.ADAPTER.decode(event))

        return signedEvent
    }

    suspend fun createPost(content: String, image: ImageManifest? = null, reference: Reference? = null): SignedEvent {
        val post = Post(content = content, image = image)

        val eventData = EventCreationData(
            content_type = ContentType.POST,
            content = Post.ADAPTER.encode(post).toByteString(),
            references = if (reference != null) listOf(reference) else emptyList(),
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createLike(subjectPointer: Pointer): SignedEvent =
        createOpinion(Opinion.LIKE, subjectPointer)

    suspend fun createDislike(subjectPointer: Pointer): SignedEvent =
        createOpinion(Opinion.DISLIKE, subjectPointer)

    suspend fun createNeutral(subjectPointer: Pointer): SignedEvent =
        createOpinion(Opinion.NEUTRAL, subjectPointer)

    suspend fun createUsername(username: String): SignedEvent {
        val lwwElement = LWWElement(
            value_ = username.encodeToByteArray().toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.USERNAME,
            lww_element = lwwElement,
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createDescription(description: String): SignedEvent {
        val lwwElement = LWWElement(
            value_ = description.encodeToByteArray().toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.DESCRIPTION,
            lww_element = lwwElement,
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createAvatar(avatar: ImageManifest): SignedEvent {
        val lwwElement = LWWElement(
            value_ = ImageManifest.ADAPTER.encode(avatar).toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.AVATAR,
            lww_element = lwwElement,
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createBanner(banner: ImageManifest): SignedEvent {
        val lwwElement = LWWElement(
            value_ = ImageManifest.ADAPTER.encode(banner).toByteString(),
            unix_milliseconds = System.currentTimeMillis(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.BANNER,
            lww_element = lwwElement,
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createFollow(system: PublicKey): SignedEvent =
        createLWWElementSetEvent(ContentType.FOLLOW, PublicKey.ADAPTER.encode(system), LWWElementSet.Operation.ADD)

    suspend fun createUnfollow(system: PublicKey): SignedEvent =
        createLWWElementSetEvent(ContentType.FOLLOW, PublicKey.ADAPTER.encode(system), LWWElementSet.Operation.REMOVE)

    suspend fun createBlock(system: PublicKey): SignedEvent =
        createLWWElementSetEvent(ContentType.BLOCK, PublicKey.ADAPTER.encode(system), LWWElementSet.Operation.ADD)

    suspend fun createUnblock(system: PublicKey): SignedEvent =
        createLWWElementSetEvent(ContentType.BLOCK, PublicKey.ADAPTER.encode(system), LWWElementSet.Operation.REMOVE)

    suspend fun createAddServer(server: String): SignedEvent =
        createLWWElementSetEvent(ContentType.SERVER, server.encodeToByteArray(), LWWElementSet.Operation.ADD)

    suspend fun createRemoveServer(server: String): SignedEvent =
        createLWWElementSetEvent(ContentType.SERVER, server.encodeToByteArray(), LWWElementSet.Operation.REMOVE)

    suspend fun createAddAuthority(authority: String): SignedEvent =
        createLWWElementSetEvent(ContentType.AUTHORITY, authority.encodeToByteArray(), LWWElementSet.Operation.ADD)

    suspend fun createRemoveAuthority(authority: String): SignedEvent =
        createLWWElementSetEvent(ContentType.AUTHORITY, authority.encodeToByteArray(), LWWElementSet.Operation.REMOVE)

    suspend fun createJoinTopic(topic: String): SignedEvent =
        createLWWElementSetEvent(ContentType.JOIN_TOPIC, topic.encodeToByteArray(), LWWElementSet.Operation.ADD)

    suspend fun createLeaveTopic(topic: String): SignedEvent =
        createLWWElementSetEvent(ContentType.JOIN_TOPIC, topic.encodeToByteArray(), LWWElementSet.Operation.REMOVE)

    suspend fun createClaim(claimType: Long, fields: List<ClaimFieldEntry>): SignedEvent {
        val claim = Claim(
            claim_type = claimType,
            fields = fields,
        )

        val eventData = EventCreationData(
            content_type = ContentType.CLAIM,
            content = Claim.ADAPTER.encode(claim).toByteString(),
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun createVerifyClaim(targetPointer: Pointer): SignedEvent {
        val targetReference = Reference(
            reference_type = 0L,
            reference = Pointer.ADAPTER.encode(targetPointer).toByteString(),
        )

        val eventData = EventCreationData(
            content_type = ContentType.VOUCH,
            references = listOf(targetReference),
            system = PublicKey(
                key_type = client.currentIdentity.keyPair.keyType,
                key = client.currentIdentity.keyPair.publicKey.key,
            ),
            process = Process(
                process = client.process!!.process,
            ),
        )

        return createEvent(eventData)
    }

    suspend fun deletePost(postPointer: Pointer): SignedEvent =
        createDelete(postPointer, ContentType.POST)
}
